packages/backend/src/core/EmailService.ts

```diff
@@ -208,19 +244,14 @@ export class EmailService {
 				blacklist: 'blacklist',
 			};
 
-			return {
-				available: false,
-				reason: validated.reason ? formatReason[validated.reason] ?? null : null,
-			};
-		}
-
-		const emailDomain: string = emailAddress.split('@')[1];
-		const isBanned = this.utilityService.isBlockedHost(this.meta.bannedEmailDomains, emailDomain);
+			// 自動追加が有効な場合はブラックリストに追加する
+			if (this.meta.enableAutoAddBannedEmailDomain) {
+				await this.addBlockedHost(emailDomain);
```

Comment:
(別の件でプライベート メールから) PR 全体を詳しく見てみましたが、これを使用して悪意を持って TLD 全体をブロックすることは可能でしょうか?  ブロックする前にメールドメインが実際に eTLD であることを確認していないため、予期しない動作を引き起こす可能性があります。

ブロック チェックの実装は次のとおりです:

https://github.com/misskey-dev/misskey/blob/develop/packages/backend/src/core/UtilityService.ts#L42-L46

これは有効なメール アドレスですが、誰かが検証で使用しようとした場合 (ただタイプミスで意図的ではない可能性もあります)、おそらく `.com` TLD 全体がブロックされます:

```js
const x = "gmail@com";

/^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/.test(x)

true
```

一般的に偶発的なブロックに対するフェイルセーフとして（完全に軽減するのは難しいと思います）、または無効だったが現在は無効ではないドメインを考慮すると、自動ブロックに自動有効期限を設定するのが良いと思います。

繰り返しの（悪意のあるか否かを問わず）リクエストを減らすために、個人的には 8 ～ 24 時間が適切であると考えます。 たとえば 4 時間後に再試行することをユーザーに知らせるもいいと思います。

---

(Branching from a private email for a different matter) I took a look closer at the entire PR, is it possible that this can be used to block whole TLDs maliciously? We did not verify that the email domain is actually an eTLD before blocking it and I think it could lead to unexpected behavior:

this is the implementation of block checking:

https://github.com/misskey-dev/misskey/blob/develop/packages/backend/src/core/UtilityService.ts#L42-L46

This is a valid email address that I can see a legitimate user typo might cause but will probably block the whole `.com` TLD if used in validation:

```js
const x = "gmail@com";

/^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/.test(x)

true
```

As a fail-safe for accidental blocks in general (which I feel is hard to completely mitigate) and considering domains that might once be invalid but not now, it might be good to set an auto-expiration for all auto-blocks. For the purpose of reducing repeated (malicous or not) requests, I personally think 8-24 hours might be appropriate.

Reply:
>悪意を持って TLD 全体をブロック

これはisBlockedHost関数で無差別にendsWithを使用していることに問題があるように思います。 
サブドメインを含めてブロックするため以下のような実装をしていると考えますが、予期しない動作を防ぐため、より制御可能な実装にすべきと思います。
```
	@bindThis
	public isBlockedHost(blockedHosts: string[], host: string | null): boolean {
		if (host == null) return false;
		return blockedHosts.some(x => `.${host.toLowerCase()}`.endsWith(`.${x}`));
	}
```

これに対する修正を先ほどPushしました。
ブラックリストが`.`から始まる場合のみ後方一致とし、それ以外の場合は完全一致にしました。
この変更は現在の運用に影響を与える可能性があるためCHANGELOG.mdに記載しました。

---

>一般的に偶発的なブロック

今まではActive Email Validationが有効でブロックされた場合は別のメールアドレスを使用するか管理者がアカウントを作成する必要がありました。
このPRでホワイトリストが追加されるため、そのような場合はホワイトリストに追加することによって問題を軽減することができると思います。